### ISSUE TITLE
[Design Choice Discussion] Instancers and Garbage Collection

### ISSUE THREAD

**Issue by soraphis** [unity-atoms/unity-atoms#370]
https://discord.com/channels/640589221136171020/643165315546742785/1000332183057203210

Scriptable Objects (even the in-scene in-memory runtime copies) are not free'd if there is no reference pointing to it like normal objects (no garbage collection).

Freeing them requires to call `Resources.UnloadUnusedAssets` which is a kinda performance heavy call.

Also it is not really possible to Destroy the atom-instances in the instancers OnDisable since there could be remaining references (if destroy works at all for them).

When fiddling around with my own SOArchitecture approach my first implementation of Instancers wouldn't create SOs, but simple classes. at runtime there is no real need for an SO. Not sure if this would be suitable for Atoms without an rework, though.

**Comment by toasterhead-master** [comment]
Yeah, that means I copy pasted a lot of code. But my hands were tied and introducing a new class to be created for each instancer won't make the code much cleaner since SO and a separate class cannot inherit the same class. And Interface default method implementations were introduced only recently with newer Unity versions, and there are still lots of c# features missing in Unity. So yeah...

**Comment by soraphis** [comment]
> Instead of creating a class like you suggested, I just made the instancer to behave exactly like an Atom would.

This has a different side effect, though.

GO1 is spawned from a prefab, and the instancer clones SO1 -> SO1_1. GO1 has a component that put's a reference of SO1_1 into a list of some manager class. GO1 is destroyed.

Currently, the reference to SO1_1 would just exist further and could be used within that list. If instead of SO1_1 the instancer is put into the list, the reference would be destroyed.

I'm not saying that we must support that scenario, just saying this is one case where backwards compatibility would be broken, whereas creating a POCO and letting the GC handle the cleanup would handle this scenario.

**Comment by toasterhead-master** [comment]
Well, in such case I guess having an additional class wouldn't hurt. Fortunately, I've implemented a unit testing infrastructure for UnityAtoms, so it will be easy to test it (I used the almighty powers of _Reflections_ to make unit testing easy with unity atoms).

**Comment by soraphis** [comment]
I personally would be totally fine with having the instancers just behave like the atom instance. Like you did. Especially if it's the easier approach.

For Ions I went full interfaces. I have PureIons (which are just classes), IonComponents (which are MonoBehaviours) and IonSOs (ScriptableObject) which all implement the same interfaces. 

a IonReference is basically just a fancy [InterfaceHelper](https://gist.github.com/soraphis/e34abebbc939e8c8a1b76f66cc91d83a) that allows putting anything that implements that interface.

Unity 2022+ definitely supports generics, to a pretty far extend just not for SerializedReferences and Assets.

**Comment by soraphis** [comment]
I personally would be totally fine wiht having the instancers just behave like the atom instance. Like you did. Especially if it's the easier approach.

For Ions I went full interfaces. I have PureIons (which are just classes), IonComponents (which are MonoBehaviours) and IonSOs (ScriptableObject) which all implement the same interfaces. 

a IonReference is basically just a fancy [InterfaceHelper](https://gist.github.com/soraphis/e34abebbc939e8c8a1b76f66cc91d83a) that allows putting anything that implements that interface.

Unity 2022+ definitely supports generics, to a pretty far extend (ScriptableReferences and as MonoBehaviors/ScriptableObject you can't instantiate them. But as a field it should actually work)

**Comment by soraphis** [comment]
My plan 4.7 is to actually change the instancers to be POCO underneath instead of in-memory assets
